app/models/api_user.rb

```diff
@@ -10,17 +18,43 @@ class ApiUser < ApplicationRecord
 
   scope :active, -> { where(revoked_at: nil) }
   scope :revoked, -> { where.not(revoked_at: nil) }
+  scope :by_name, -> { reorder(:name) }
+
+  validates :name, presence: true, uniqueness: {scope: :local_authority_id, conditions: -> { where(revoked_at: nil) }}
+  validates :token, format: {with: TOKEN_FORMAT}, on: :create
+  validates :file_downloader, store_model: true
 
-  validates :name, presence: true, uniqueness: {scope: :local_authority_id}
-  validates :file_downloader, store_model: {allow_blank: true}
+  accepts_nested_attributes_for :file_downloader
 
   class << self
     def authenticate(token)
-      active.find_by(token: token)
+      active.find_by(token: token).tap { |user| user.try(:touch, :last_used_at) }
+    end
+
+    def generate_unique_secure_token(length: 36)
+      super.then { |token| "bops_#{token}#{checksum(token)}" }
```

Comment:
GitHub suggests adding the checksum so you can validate it without making a request but unless I'm missing something an attacker could generate the valid checksum anyway. Better option maybe to create a signed secret value that we can change the key to and invalidate all tokens if there were to be a mass leak.

Reply:
Yeah I think you're right, it would in theory allow us to check that the _format_ was valid but that's hardly worth the saving — unless I'm missing something. (It's not like ISBNs or credit card numbers where checking for misentered numbers might be useful.)